In [1]:
#import libraries
import pandas as pd
import numpy as np

In [2]:
# Load data into a dataframe which called as 'match'

match = pd.read_csv('D:/Self Study/Predict football match winners with machine learning/Data/matches.csv',index_col=0)
# match = pd.read_csv('../Data/matches.csv',index_col=0)


# read first five rows

match.head()

date   time            comp        round  day venue result   gf   ga  \
1  2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2  2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3  2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4  2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6  2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...  match report  notes    sh   sot  dist   fk   pk pkatt  \
1       Tottenham  ...  Match Report    NaN  18.0   4.0  16.9  1.0  0.0   0.0   
2    Norwich City  ...  Match Report    NaN  16.0   4.0  17.3  1.0  0.0   0.0   
3         Arsenal  ...  Match Report    NaN  25.0  10.0  14.3  0.0  0.0   0.0   
4  Leicester City  ...  Match Report    NaN  25.0   8.0  14.0  0.0  0.0   0.0   
6     Southampton  ...  Match Report    NaN  16.0   1.0  15.7  1.0  0.0   0.0   

   season             team  
1    2022  Manchester City  
2    2022  Manchester City  
3    2022  Manchester City  
4    2022  Manchester City  
6    2022  Manchester City  

[5 rows x 27 columns]

In [3]:
# check the shape of the dataframe
match.shape

(1389, 27)

There can be seen 1389 rows and 27 columns in this dataset.

# Investigating missing data

In [4]:
# check missing values
match.info()


<class 'pandas.core.frame.DataFrame'>
Int64Index: 1389 entries, 1 to 42
Data columns (total 27 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   date          1389 non-null   object 
 1   time          1389 non-null   object 
 2   comp          1389 non-null   object 
 3   round         1389 non-null   object 
 4   day           1389 non-null   object 
 5   venue         1389 non-null   object 
 6   result        1389 non-null   object 
 7   gf            1389 non-null   float64
 8   ga            1389 non-null   float64
 9   opponent      1389 non-null   object 
 10  xg            1389 non-null   float64
 11  xga           1389 non-null   float64
 12  poss          1389 non-null   float64
 13  attendance    693 non-null    float64
 14  captain       1389 non-null   object 
 15  formation     1389 non-null   object 
 16  referee       1389 non-null   object 
 17  match report  1389 non-null   object 
 18  notes         0 non-null      

There can be seen attendence,notes columns have a few missing data. but we can not get a clue about the all the columns by using this code.


###### In EPL Matches, 38 matches playing in each season, 20 teams will participate on it. In this dataset we are having 2 seasons which are 2020-2021 season and 2021-2022 season.So,
###### the total number of matches  = (no of matches in a season * no of teams * no of seasons)
######                              = 38 * 20 * 2
######                              =1520

But we actually have 1389 matches. Therefore, Need to find a reason for this. Let's look at the number of matches played by each team.

In [5]:
match['team'].value_counts()

Southampton                 72
Brighton and Hove Albion    72
Manchester United           72
West Ham United             72
Newcastle United            72
Burnley                     71
Leeds United                71
Crystal Palace              71
Manchester City             71
Wolverhampton Wanderers     71
Tottenham Hotspur           71
Arsenal                     71
Leicester City              70
Chelsea                     70
Aston Villa                 70
Everton                     70
Liverpool                   38
Fulham                      38
West Bromwich Albion        38
Sheffield United            38
Brentford                   34
Watford                     33
Norwich City                33
Name: team, dtype: int64

Most of the teams have match count over 70 and we have a few teams which are having match count 30.

In EPL,
- three teams are relegated in each year, they'are demoted to the lower league ,and 
- three teams are pulled up from the lower league into the EPL.

So, we would expect six teams to have fewer than 70 matches in our dataset.But we actually have seven teams that have fewer than 70 matches.

- Important : The team "liverpool" was not relegated last season.
- So, let's check our macthes from 'liverpool'.

In [6]:
match[match["team"] == "Liverpool"]

date   time            comp         round  day venue result   gf  \
1   2020-09-12  17:30  Premier League   Matchweek 1  Sat  Home      W  4.0   
2   2020-09-20  16:30  Premier League   Matchweek 2  Sun  Away      W  2.0   
4   2020-09-28  20:00  Premier League   Matchweek 3  Mon  Home      W  3.0   
6   2020-10-04  19:15  Premier League   Matchweek 4  Sun  Away      L  2.0   
7   2020-10-17  12:30  Premier League   Matchweek 5  Sat  Away      D  2.0   
9   2020-10-24  20:00  Premier League   Matchweek 6  Sat  Home      W  2.0   
11  2020-10-31  17:30  Premier League   Matchweek 7  Sat  Home      W  2.0   
13  2020-11-08  16:30  Premier League   Matchweek 8  Sun  Away      D  1.0   
14  2020-11-22  19:15  Premier League   Matchweek 9  Sun  Home      W  3.0   
16  2020-11-28  12:30  Premier League  Matchweek 10  Sat  Away      D  1.0   
18  2020-12-06  19:15  Premier League  Matchweek 11  Sun  Home      W  4.0   
20  2020-12-13  16:30  Premier League  Matchweek 12  Sun  Away      D  1.0   
21  2020-12-16  20:00  Premier League  Matchweek 13  Wed  Home      W  2.0   
22  2020-12-19  12:30  Premier League  Matchweek 14  Sat  Away      W  7.0   
23  2020-12-27  16:30  Premier League  Matchweek 15  Sun  Home      D  1.0   
24  2020-12-30  20:00  Premier League  Matchweek 16  Wed  Away      D  0.0   
25  2021-01-04  20:00  Premier League  Matchweek 17  Mon  Away      L  0.0   
27  2021-01-17  16:30  Premier League  Matchweek 19  Sun  Home      D  0.0   
28  2021-01-21  20:00  Premier League  Matchweek 18  Thu  Home      L  0.0   
30  2021-01-28  20:00  Premier League  Matchweek 20  Thu  Away      W  3.0   
31  2021-01-31  16:30  Premier League  Matchweek 21  Sun  Away      W  3.0   
32  2021-02-03  20:15  Premier League  Matchweek 22  Wed  Home      L  0.0   
33  2021-02-07  16:30  Premier League  Matchweek 23  Sun  Home      L  1.0   
34  2021-02-13  12:30  Premier League  Matchweek 24  Sat  Away      L  1.0   
36  2021-02-20  17:30  Premier League  Matchweek 25  Sat  Home      L  0.0   
37  2021-02-28  19:15  Premier League  Matchweek 26  Sun  Away      W  2.0   
38  2021-03-04  20:15  Premier League  Matchweek 29  Thu  Home      L  0.0   
39  2021-03-07  14:00  Premier League  Matchweek 27  Sun  Home      L  0.0   
41  2021-03-15  20:00  Premier League  Matchweek 28  Mon  Away      W  1.0   
42  2021-04-03  20:00  Premier League  Matchweek 30  Sat  Away      W  3.0   
44  2021-04-10  15:00  Premier League  Matchweek 31  Sat  Home      W  2.0   
46  2021-04-19  20:00  Premier League  Matchweek 32  Mon  Away      D  1.0   
47  2021-04-24  12:30  Premier League  Matchweek 33  Sat  Home      D  1.0   
48  2021-05-08  20:15  Premier League  Matchweek 35  Sat  Home      W  2.0   
49  2021-05-13  20:15  Premier League  Matchweek 34  Thu  Away      W  4.0   
50  2021-05-16  16:30  Premier League  Matchweek 36  Sun  Away      W  2.0   
51  2021-05-19  20:15  Premier League  Matchweek 37  Wed  Away      W  3.0   
52  2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  2.0   

     ga         opponent  ...  match report  notes    sh   sot  dist   fk  \
1   3.0     Leeds United  ...  Match Report    NaN  20.0   4.0  17.0  0.0   
2   0.0          Chelsea  ...  Match Report    NaN  17.0   5.0  17.7  1.0   
4   1.0          Arsenal  ...  Match Report    NaN  21.0   9.0  16.8  0.0   
6   7.0      Aston Villa  ...  Match Report    NaN  14.0   8.0  15.8  1.0   
7   2.0          Everton  ...  Match Report    NaN  22.0   8.0  15.0  1.0   
9   1.0    Sheffield Utd  ...  Match Report    NaN  17.0   5.0  18.2  1.0   
11  1.0         West Ham  ...  Match Report    NaN   8.0   2.0  18.6  1.0   
13  1.0  Manchester City  ...  Match Report    NaN   9.0   2.0  21.5  0.0   
14  0.0   Leicester City  ...  Match Report    NaN  24.0  12.0  11.9  0.0   
16  1.0         Brighton  ...  Match Report    NaN   6.0   2.0  20.9  0.0   
18  0.0           Wolves  ...  Match Report    NaN  11.0   6.0  16.6  1.0   
20  1.0           Fulham  ...  Match Report

- We can see here we have only one data from 2020 - 2021 season for liverpool.
- We are missing data from the next season.
- So now we know where some of these missing rows went.
- We're missing one season for liverpool.


Let's look at the match data."round" columns tells us which match week each match was played.

In [7]:
#check match data
match["round"].value_counts()

Matchweek 1     39
Matchweek 16    39
Matchweek 34    39
Matchweek 32    39
Matchweek 31    39
Matchweek 29    39
Matchweek 28    39
Matchweek 26    39
Matchweek 25    39
Matchweek 24    39
Matchweek 23    39
Matchweek 2     39
Matchweek 19    39
Matchweek 17    39
Matchweek 20    39
Matchweek 15    39
Matchweek 5     39
Matchweek 3     39
Matchweek 13    39
Matchweek 12    39
Matchweek 4     39
Matchweek 11    39
Matchweek 10    39
Matchweek 9     39
Matchweek 8     39
Matchweek 14    39
Matchweek 7     39
Matchweek 6     39
Matchweek 30    37
Matchweek 27    37
Matchweek 22    37
Matchweek 21    37
Matchweek 18    37
Matchweek 33    32
Matchweek 35    20
Matchweek 36    20
Matchweek 37    20
Matchweek 38    20
Name: round, dtype: int64

We should have '39' matches becaues we are missing "liverpool'.

There are 10 weeks that the number of matches less than 39. The reason for that is these data are scrpaed from the website and , at that time the season 2021-2022 was still ongoing.

# Data Cleaning

In [8]:
# check data types
match.dtypes

date             object
time             object
comp             object
round            object
day              object
venue            object
result           object
gf              float64
ga              float64
opponent         object
xg              float64
xga             float64
poss            float64
attendance      float64
captain          object
formation        object
referee          object
match report     object
notes           float64
sh              float64
sot             float64
dist            float64
fk              float64
pk              float64
pkatt           float64
season            int64
team             object
dtype: object

- ML algorthms can only work with Numeric data.(float 64, int 64).
- ML algorithms can't work with data that is an object.
- So we need to do figure out how to create columns that we can use as predictors.

In [9]:
# change date column datatype'object' to 'datetime'
# can extract data from like day of week, hour, day, year, month 

match ['date'] = pd.to_datetime(match['date'])

In [10]:
# check the date column data type

match.dtypes

date            datetime64[ns]
time                    object
comp                    object
round                   object
day                     object
venue                   object
result                  object
gf                     float64
ga                     float64
opponent                object
xg                     float64
xga                    float64
poss                   float64
attendance             float64
captain                 object
formation               object
referee                 object
match report            object
notes                  float64
sh                     float64
sot                    float64
dist                   float64
fk                     float64
pk                     float64
pkatt                  float64
season                   int64
team                    object
dtype: object

# Creating Predictors

1. First Predictor: 'venue'.
- need to convert 'home' or 'away' column into a numeric column.
- 'venue indicate whether the team played an 'away' game or 'home'game.
- One of the most important predictor for the outcomes of matches. ( ex: homefield advantage)
- going to convert this to numbers.


In [11]:
# There are two unique values. ('away','home')
# using ".astype" it is converting from'string' into 'category' type, and 
# then using ".cat.codes" will convert 'categories' into the 'integer' type.
# this will create a new column known as 'venue_code'
# It will contain numbers

match["venue_code"] = match["venue"].astype("category").cat.codes

In [12]:
match

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...  notes    sh   sot  dist   fk   pk pkatt season  \
1   1.0       Tottenham  ...    NaN  18.0   4.0  16.9  1.0  0.0   0.0   2022   
2   0.0    Norwich City  ...    NaN  16.0   4.0  17.3  1.0  0.0   0.0   2022   
3   0.0         Arsenal  ...    NaN  25.0  10.0  14.3  0.0  0.0   0.0   2022   
4   0.0  Leicester City  ...    NaN  25.0   8.0  14.0  0.0  0.0   0.0   2022   
6   0.0     Southampton  ...    NaN  16.0   1.0  15.7  1.0  0.0   0.0   2022   
..  ...             ...  ...    ...   ...   ...   ...  ...  ...   ...    ...   
38  4.0       Tottenham  ...    NaN   8.0   1.0  17.4  0.0  0.0   0.0   2021   
39  2.0  Crystal Palace  ...    NaN   7.0   0.0  11.4  1.0  0.0   0.0   2021   
40  0.0         Everton  ...    NaN  10.0   3.0  17.0  0.0  0.0   0.0   2021   
41  1.0   Newcastle Utd  ...    NaN  11.0   1.0  16.0  1.0  0.0   0.0   2021   
42  0.0         Burnley  ...    NaN  12.0   3.0  17.0  0.0  0.0   0.0   2021   

                team  venue_code  
1    Manchester City           0  
2    Manchester City           1  
3    Manchester City           1  
4    Manchester City           0  
6    Manchester City           1  
..               ...         ...  
38  Sheffield United           0  
39  Sheffield United           1  
40  Sheffield United           0  
41  Sheffield United           0  
42  Sheffield United           1  

[1389 rows x 28 columns]

 2. Second Predictor : 'opponent'.
 - Will follow the same method used above.

In [13]:
match["opp_code"] = match["opponent"].astype("category").cat.codes
match

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...    sh   sot  dist   fk   pk pkatt season  \
1   1.0       Tottenham  ...  18.0   4.0  16.9  1.0  0.0   0.0   2022   
2   0.0    Norwich City  ...  16.0   4.0  17.3  1.0  0.0   0.0   2022   
3   0.0         Arsenal  ...  25.0  10.0  14.3  0.0  0.0   0.0   2022   
4   0.0  Leicester City  ...  25.0   8.0  14.0  0.0  0.0   0.0   2022   
6   0.0     Southampton  ...  16.0   1.0  15.7  1.0  0.0   0.0   2022   
..  ...             ...  ...   ...   ...   ...  ...  ...   ...    ...   
38  4.0       Tottenham  ...   8.0   1.0  17.4  0.0  0.0   0.0   2021   
39  2.0  Crystal Palace  ...   7.0   0.0  11.4  1.0  0.0   0.0   2021   
40  0.0         Everton  ...  10.0   3.0  17.0  0.0  0.0   0.0   2021   
41  1.0   Newcastle Utd  ...  11.0   1.0  16.0  1.0  0.0   0.0   2021   
42  0.0         Burnley  ...  12.0   3.0  17.0  0.0  0.0   0.0   2021   

                team  venue_code  opp_code  
1    Manchester City           0        18  
2    Manchester City           1        15  
3    Manchester City           1         0  
4    Manchester City           0        10  
6    Manchester City           1        17  
..               ...         ...       ...  
38  Sheffield United           0        18  
39  Sheffield United           1         6  
40  Sheffield United           0         7  
41  Sheffield United           0        14  
42  Sheffield United           1         4  

[1389 rows x 29 columns]

In [14]:
#Here we are going to remove the colon and the minutes
# Keep the hour
# will do it with a string replacement 
# using regular expression going to repalce the colon and the minutes with nothing 

match["hour"] = match["time"].str.replace(":.+","",regex = True).astype("int")

In [15]:
# Create a new column to represent each day with a number
# ex: sunday : 6, Saturday: 5

match["day_code"] = match ["date"].dt.dayofweek

match

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
38 2021-05-02  19:15  Premier League  Matchweek 34  Sun  Away      L  0.0   
39 2021-05-08  15:00  Premier League  Matchweek 35  Sat  Home      L  0.0   
40 2021-05-16  19:00  Premier League  Matchweek 36  Sun  Away      W  1.0   
41 2021-05-19  18:00  Premier League  Matchweek 37  Wed  Away      L  0.0   
42 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  1.0   

     ga        opponent  ...  dist   fk   pk  pkatt season              team  \
1   1.0       Tottenham  ...  16.9  1.0  0.0    0.0   2022   Manchester City   
2   0.0    Norwich City  ...  17.3  1.0  0.0    0.0   2022   Manchester City   
3   0.0         Arsenal  ...  14.3  0.0  0.0    0.0   2022   Manchester City   
4   0.0  Leicester City  ...  14.0  0.0  0.0    0.0   2022   Manchester City   
6   0.0     Southampton  ...  15.7  1.0  0.0    0.0   2022   Manchester City   
..  ...             ...  ...   ...  ...  ...    ...    ...               ...   
38  4.0       Tottenham  ...  17.4  0.0  0.0    0.0   2021  Sheffield United   
39  2.0  Crystal Palace  ...  11.4  1.0  0.0    0.0   2021  Sheffield United   
40  0.0         Everton  ...  17.0  0.0  0.0    0.0   2021  Sheffield United   
41  1.0   Newcastle Utd  ...  16.0  1.0  0.0    0.0   2021  Sheffield United   
42  0.0         Burnley  ...  17.0  0.0  0.0    0.0   2021  Sheffield United   

   venue_code opp_code  hour  day_code  
1           0       18    16         6  
2           1       15    15         5  
3           1        0    12         5  
4           0       10    15         5  
6           1       17    15         5  
..        ...      ...   ...       ...  
38          0       18    19         6  
39          1        6    15         5  
40          0        7    19         6  
41          0       14    18         2  
42          1        4    16         6  

[1389 rows x 31 columns]

- Now we can start to train our initial machine learning model. 
- Before, start it we need to actually set up a 'target', that we're going to try to predict.
- So our target is actually going to be whether the team won or not.

- In our data frame in the 'result' column , it can be 'L' or 'W' or 'D'.
- So, we are going to convert this into two numbers, if the result is a, 'Loss' or a 'draw' will convert it as a '0' and if the result is 'Won' it will convert to '1'. Because what we want to predict is if the team won or not.

- Another method:
If we want to predict "Losses" and "Draws" seperately ,then we can use '0' or '1' or '2'. But it will increase the complexity of the model.

- But, in this code we are gonna combine them just to keep the simplexity.

In [16]:
# result will be a bollean value which is 'True' or 'False' 
# True for if the column includes 'W' if not it will be 'False'
# Here using .astype to convert this bollean values into numbers 
# this will turn 'True' into '1' and 'False' into '0'
match ["target"] = (match["result"] == "W").astype("int")
match.head()

date   time            comp        round  day venue result   gf   ga  \
1 2021-08-15  16:30  Premier League  Matchweek 1  Sun  Away      L  0.0  1.0   
2 2021-08-21  15:00  Premier League  Matchweek 2  Sat  Home      W  5.0  0.0   
3 2021-08-28  12:30  Premier League  Matchweek 3  Sat  Home      W  5.0  0.0   
4 2021-09-11  15:00  Premier League  Matchweek 4  Sat  Away      W  1.0  0.0   
6 2021-09-18  15:00  Premier League  Matchweek 5  Sat  Home      D  0.0  0.0   

         opponent  ...   fk   pk  pkatt  season             team venue_code  \
1       Tottenham  ...  1.0  0.0    0.0    2022  Manchester City          0   
2    Norwich City  ...  1.0  0.0    0.0    2022  Manchester City          1   
3         Arsenal  ...  0.0  0.0    0.0    2022  Manchester City          1   
4  Leicester City  ...  0.0  0.0    0.0    2022  Manchester City          0   
6     Southampton  ...  1.0  0.0    0.0    2022  Manchester City          1   

  opp_code hour  day_code  target  
1       18   16         6       0  
2       15   15         5       1  
3        0   12         5       1  
4       10   15         5       1  
6       17   15         5       0  

[5 rows x 32 columns]

# Creating Initial Machine Learning Model

In [17]:
#import  random forest classifier model form sci-kit learn
# random forest is a type of machine learning model that can pick-up non-linearities in the data

from sklearn.ensemble import RandomForestClassifier


In [18]:
# initialize the random forest classifier class

rf = RandomForestClassifier(n_estimators=50,min_samples_split= 10,random_state=1)


### Here we are going to split up the train and test data sets
- the data are time series data

- so we have to carefull when we split it to make to ensure that the all the data in the test set comes after all the data in the training set
- because in the real world we can't train an algorithm on data in the future and use it to predict data in the past
It has to be other way around you can only predict the future using the past 
- So we basically take anything that came brfore 2022.So all of the matches before 2022 ,and put those in our training set.
-  Then our test set wwill be anything in 2022

In [19]:
# create a subset of match dataframe for training data
train = match[match['date'] < '2022-01-01']

In [20]:
# create a subset of match dataset for test data
test = match [match ['date'] > '2022- 01-01 ']

In [21]:
# cretae predictors
# this is a list of predictor columns
predictors = ["venue_code","opp_code","hour","day_code"]

In [22]:
# fit random forest model
# we will fit it using train predictors
# we will try to predict our target
rf.fit(train[predictors],train["target"])

RandomForestClassifier(min_samples_split=10, n_estimators=50, random_state=1)

In [23]:
# Generate predictions 
preds = rf.predict(test[predictors])

Next we need to figure out how we can actually figure out how we determine the accuracy of the model. So we are gonna try a couple of different metrics and see which one makes more sense.
- We'll import from scikitlearn.metrics will import the 'accuracy_score'. 
- So 'accuracy_score' is a metric that will basically say,

    - if you predicted a 'win' what percentage of the time did the team 'actually win' 
    - if you predicted a 'loss' what percentage of the time did the team 'actually lose'
    
     and it'll combine that basically looking at what percentage of the time was your prediction accurate.

In [24]:
# import accuracy_score from scikit learn
from sklearn.metrics import accuracy_score

In [25]:
# Here we will pass the actual (test["target"]) and predictions (pred)
acc =accuracy_score(test["target"],preds)

In [26]:
# check the acuuaracy
acc

0.6123188405797102

We can see our accuracy looks pretty good. So when we predicted something would happen 61 of the time that thing happend.

Let's dig a liitle bit deeper into this though and see in which situations our acuuracy was high versus low.
In order to do that first we need to create a dataframe.

In [27]:
# create a dataframe to combine our actual values and predicted values
combined = pd.DataFrame(dict(actual = test ["target"],prediction= preds))

In [28]:
# create a cross tab
# this is just a two way table that will show us when we predicted a zero or a one what actually happened 
pd.crosstab(index = combined["actual"], columns = combined["prediction"])

prediction    0   1
actual             
0           141  31
1            76  28

When we predict a 'loss' or a 'draw' we can see that most of the time we were correct.
- '141' times we were correct only '76' times we were wrong where the actual  'was a win'.
- But when we 'preicted a win' we were actually wrong more often than we were right.

So since what we care about this model is 'predicting wins'. we need to go back and revise our accuracy metric a little bit.
So we're actually going to use a slightly different metric.

- We'are actually going to use something called the 'precision score'.
    - This is gonna tell us when we predicted a 'win' what percentage of the time did the team 'actually win'.

In [29]:
# import precision_score

from sklearn.metrics import precision_score

# pass actual values and prediction values

precision_score(test["target"], preds)

0.4745762711864407

- Here you can see that precision was only 47%.
- So when we 'predicted a win' the team only 'actually won 47 percent of the time' which is not great precision.
- So we can improve this model a quite a bit and we will do that in the next step.


# Improving precision with Rolling Averages

- Next we are gonna create more predictors to improve the accuracy of the model. 
- To do this we're going to need to split our 'match' dataframe up by 'team'.Because what we want to do is compute "Rolling Averages" for 'team performance'. 
    - How did this team do in the past few games?
    - How many shots did they have ?
    - How many goals were scored againest them? etc.

So we are gonna create a new dataframe called as 'grouped_matches'

In [30]:
# Create new dataframe which is called as 'grouped_matches'
# this will create one dataframe for every squad in our data 

grouped_matches = match.groupby("team")
# grouped_matches.head()

In [31]:
# look at a single group

group = grouped_matches.get_group("Manchester City")

In [32]:
group

date   time            comp         round  day venue result   gf  \
1  2021-08-15  16:30  Premier League   Matchweek 1  Sun  Away      L  0.0   
2  2021-08-21  15:00  Premier League   Matchweek 2  Sat  Home      W  5.0   
3  2021-08-28  12:30  Premier League   Matchweek 3  Sat  Home      W  5.0   
4  2021-09-11  15:00  Premier League   Matchweek 4  Sat  Away      W  1.0   
6  2021-09-18  15:00  Premier League   Matchweek 5  Sat  Home      D  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
54 2021-05-01  12:30  Premier League  Matchweek 34  Sat  Away      W  2.0   
56 2021-05-08  17:30  Premier League  Matchweek 35  Sat  Home      L  1.0   
57 2021-05-14  20:00  Premier League  Matchweek 36  Fri  Away      W  4.0   
58 2021-05-18  19:00  Premier League  Matchweek 37  Tue  Away      L  2.0   
59 2021-05-23  16:00  Premier League  Matchweek 38  Sun  Home      W  5.0   

     ga        opponent  ...   fk   pk  pkatt  season             team  \
1   1.0       Tottenham  ...  1.0  0.0    0.0    2022  Manchester City   
2   0.0    Norwich City  ...  1.0  0.0    0.0    2022  Manchester City   
3   0.0         Arsenal  ...  0.0  0.0    0.0    2022  Manchester City   
4   0.0  Leicester City  ...  0.0  0.0    0.0    2022  Manchester City   
6   0.0     Southampton  ...  1.0  0.0    0.0    2022  Manchester City   
..  ...             ...  ...  ...  ...    ...     ...              ...   
54  0.0  Crystal Palace  ...  1.0  0.0    0.0    2021  Manchester City   
56  2.0         Chelsea  ...  0.0  0.0    1.0    2021  Manchester City   
57  3.0   Newcastle Utd  ...  1.0  0.0    0.0    2021  Manchester City   
58  3.0        Brighton  ...  1.0  0.0    0.0    2021  Manchester City   
59  0.0         Everton  ...  0.0  0.0    0.0    2021  Manchester City   

   venue_code opp_code hour  day_code  target  
1           0       18   16         6       0  
2           1       15   15         5       1  
3           1        0   12         5       1  
4           0       10   15         5       1  
6           1       17   15         5       0  
..        ...      ...  ...       ...     ...  
54          0        6   12         5       1  
56          1        5   17         5       0  
57          0       14   20         4       1  
58          0        3   19         1       0  
59          1        7   16         6       1  

[71 rows x 32 columns]

- This will give us a 'single group' from the data and this is just 'Manchester city'. So these are all of the matches that manchester city played in.
- So what we wanna do is compute "rolling avaerages".So what we want to say is if we are 'match week 4' how did 'manchester city' do in the 'previous three match weeks' and then we can pass that into the algorithm and the algorithm can use that information to make better predictions about how they'll do in 'match week 4'.

Ex: if they lost "all three of their games" before "match week four" you might expect them to also "lose match week four" because they haven't been playing very well.

 So we are gonna create a function which named as "rolling_averages".This will do is,
 - take a group in (group)
 - it'll take a set of columns (cols) that we want to compute rolling averages for.
 - and then it will take in a set of new columns that we want to assign the rolling averages to. (new_cols)



In [33]:
#create a function which named as 'rolling averages' 



def rolling_averages(group,cols,new_cols):
    # sort the 'group' by date
    # We want it to be sorted in ascending order of date
    # We want to look at the last three matches a team played and what their performance was like

    group = group.sort_values("date")

    # create a variable calling 'rolling stat
    # what we are gonna do is take a set of columns which will be in this column list 
    # compute the rolling averages for those columns
    # closed ='left' parameter is very important
    rolling_stats = group[cols].rolling(3,closed = 'left').mean()
    
    # assign 'rolling_stats' values back to original dataframe
    group[new_cols] = rolling_stats

    # drop missing values
    group = group.dropna(subset =new_cols)

    return group

Column name definitions :
- gf - goals for
- ga - goals against
- sh - shots 
- sot - shots on target
- dist - distance (the distance each shot travelled)
- fk - free kicks
- pk - penalty kicks
- pkatt - penalty kick attempts

In [34]:
# create a list of columns that we want to compute rolling averages
cols = ["gf","ga","sh","sot","dist","fk","pk","pkatt"]



In [35]:
# create new column list which is just going to be a format 'string'
new_cols = [f"{c}_rolling" for c in cols]

new_cols

['gf_rolling',
 'ga_rolling',
 'sh_rolling',
 'sot_rolling',
 'dist_rolling',
 'fk_rolling',
 'pk_rolling',
 'pkatt_rolling']

Now we can call rolling averages for a single group and our group is manchester city. So this is just for the manchester city games.  

In [36]:
rolling_averages(group,cols,new_cols)

# rolling_averages

date   time            comp         round  day venue result   gf  \
5  2020-10-17  17:30  Premier League   Matchweek 5  Sat  Home      W  1.0   
7  2020-10-24  12:30  Premier League   Matchweek 6  Sat  Away      D  1.0   
9  2020-10-31  12:30  Premier League   Matchweek 7  Sat  Away      W  1.0   
11 2020-11-08  16:30  Premier League   Matchweek 8  Sun  Home      D  1.0   
12 2020-11-21  17:30  Premier League   Matchweek 9  Sat  Away      L  0.0   
..        ...    ...             ...           ...  ...   ...    ...  ...   
42 2022-03-14  20:00  Premier League  Matchweek 29  Mon  Away      D  0.0   
44 2022-04-02  15:00  Premier League  Matchweek 31  Sat  Away      W  2.0   
46 2022-04-10  16:30  Premier League  Matchweek 32  Sun  Home      D  2.0   
49 2022-04-20  20:00  Premier League  Matchweek 30  Wed  Home      W  3.0   
50 2022-04-23  15:00  Premier League  Matchweek 34  Sat  Home      W  5.0   

     ga        opponent  ...  day_code  target  gf_rolling  ga_rolling  \
5   0.0         Arsenal  ...         5       1    2.000000    2.333333   
7   1.0        West Ham  ...         5       0    1.333333    2.000000   
9   0.0   Sheffield Utd  ...         5       1    1.000000    0.666667   
11  1.0       Liverpool  ...         6       0    1.000000    0.333333   
12  2.0       Tottenham  ...         5       0    1.000000    0.666667   
..  ...             ...  ...       ...     ...         ...         ...   
42  0.0  Crystal Palace  ...         0       0    2.333333    1.333333   
44  0.0         Burnley  ...         5       1    1.666667    0.333333   
46  2.0       Liverpool  ...         6       0    2.000000    0.333333   
49  0.0        Brighton  ...         2       1    1.333333    0.666667   
50  1.0         Watford  ...         5       1    2.333333    0.666667   

   sh_rolling sot_rolling dist_rolling fk_rolling  pk_rolling  pkatt_rolling  
5   17.333333    4.666667    18.900000   1.333333    0.333333       0.333333  
7   17.333333    3.666667    17.733333   0.666667    0.000000       0.000000  
9   16.666667    4.333333    18.233333   0.666667    0.000000       0.000000  
11  14.333333    6.666667    18.466667   1.000000    0.000000       0.000000  
12  12.000000    5.666667    19.366667   1.000000    0.000000       0.333333  
..        ...         ...          ...        ...         ...            ...  
42  19.000000    7.000000    15.366667   0.333333    0.333333       0.333333  
44  18.333333    7.333333    16.000000   0.333333    0.000000       0.000000  
46  20.000000    6.666667    16.133333   0.333333    0.000000       0.000000  
49  15.666667    4.666667    16.700000   0.333333    0.000000       0.000000  
50  15.333333    5.000000    17.200000   0.333333    0.000000       0.000000  

[68 rows x 40 columns]

We can apply this to all of our matches

In [37]:
macth_rolling = match.groupby("team").apply(lambda x: rolling_averages(x,cols,new_cols))

In [38]:
macth_rolling

date   time            comp         round  \
team                                                                         
Arsenal                 6  2020-10-04  14:00  Premier League   Matchweek 4   
                        7  2020-10-17  17:30  Premier League   Matchweek 5   
                        9  2020-10-25  19:15  Premier League   Matchweek 6   
                        11 2020-11-01  16:30  Premier League   Matchweek 7   
                        13 2020-11-08  19:15  Premier League   Matchweek 8   
...                               ...    ...             ...           ...   
Wolverhampton Wanderers 32 2022-03-13  14:00  Premier League  Matchweek 29   
                        33 2022-03-18  20:00  Premier League  Matchweek 30   
                        34 2022-04-02  15:00  Premier League  Matchweek 31   
                        35 2022-04-08  20:00  Premier League  Matchweek 32   
                        36 2022-04-24  14:00  Premier League  Matchweek 34   

                            day venue result   gf   ga         opponent  ...  \
team                                                                     ...   
Arsenal                 6   Sun  Home      W  2.0  1.0    Sheffield Utd  ...   
                        7   Sat  Away      L  0.0  1.0  Manchester City  ...   
                        9   Sun  Home      L  0.0  1.0   Leicester City  ...   
                        11  Sun  Away      W  1.0  0.0   Manchester Utd  ...   
                        13  Sun  Home      L  0.0  3.0      Aston Villa  ...   
...                         ...   ...    ...  ...  ...              ...  ...   
Wolverhampton Wanderers 32  Sun  Away      W  1.0  0.0          Everton  ...   
                        33  Fri  Home      L  2.0  3.0     Leeds United  ...   
                        34  Sat  Home      W  2.0  1.0      Aston Villa  ...   
                        35  Fri  Away      L  0.0  1.0    Newcastle Utd  ...   
                        36  Sun  Away      L  0.0  1.0          Burnley  ...   

                            day_code  target  gf_rolling  ga_rolling  \
team                                                                   
Arsenal                 6          6       1    2.000000    1.333333   
                        7          5       0    1.666667    1.666667   
                        9          6       0    1.000000    1.666667   
                        11         6       1    0.666667    1.000000   
                        13         6       0    0.333333    0.666667   
...                              ...     ...         ...         ...   
Wolverhampton Wanderers 32         6       1    1.333333    1.000000   
                        33         4       0    1.666667    0.666667   
                        34         5       1    2.333333    1.000000   
                        35         4       0    1.666667    1.333333   
                        36         6       0    1.333333    1.666667   

                           sh_rolling sot_rolling dist_rolling fk_rolling  \
team                                                                        
Arsenal                 6    7.666667    3.666667    14.733333   0.666667   
                        7    5.333333    3.666667    15.766667   0.000000   
                        9    7.000000    3.666667    16.733333   0.666667   
                        11   9.666667    4.000000    16.033333   1.000000   
                        13   9.666667    2.666667    18.033333   1.000000   
...                               ...         ...          ...        ...   
Wolverhampton Wanderers 32  12.333333    3.666667    19.300000   0.000000   
                        33  12.333333    4.333333    19.600000   0.000000   
                        34  13.000000    5.333333    19.833333   0.000000   
                        35  13.000000    5.000000    18.533333   0.000000   
                        36  10.000000    4.666667    17.633333   0.000000   

                            pk_rolling